lib/codecs/src/encoding/format/protobuf.rs

```diff
@@ -0,0 +1,342 @@
+use crate::encoding::BuildError;
+use bytes::BytesMut;
+use prost::Message;
+use prost_reflect::{DescriptorPool, DynamicMessage, FieldDescriptor, Kind, MessageDescriptor};
+use std::path::{Path, PathBuf};
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+fn get_message_descriptor(
+    descriptor_set_path: &Path,
+    message_type: &str,
+) -> vector_common::Result<MessageDescriptor> {
+    let b = std::fs::read(descriptor_set_path).map_err(|e| {
+        format!("Failed to open protobuf desc file '{descriptor_set_path:?}': {e}",)
+    })?;
+    let pool = DescriptorPool::decode(b.as_slice()).map_err(|e| {
+        format!("Failed to parse protobuf desc file '{descriptor_set_path:?}': {e}")
+    })?;
+    pool.get_message_by_name(message_type).ok_or_else(|| {
+        format!("The message type '{message_type}' could not be found in '{descriptor_set_path:?}'")
+            .into()
+    })
+}
+
+/// Config used to build a `ProtobufSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct ProtobufSerializerConfig {
+    /// Options for the Protobuf serializer.
+    pub protobuf: ProtobufSerializerOptions,
+}
+
+impl ProtobufSerializerConfig {
+    /// Build the `ProtobufSerializer` from this configuration.
+    pub fn build(&self) -> Result<ProtobufSerializer, BuildError> {
+        let message_descriptor = get_message_descriptor(
+            &self.protobuf.descriptor_set_path,
+            &self.protobuf.message_type,
+        )?;
+        Ok(ProtobufSerializer { message_descriptor })
+    }
+
+    /// The data type of events that are accepted by `ProtobufSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log.and(DataType::Trace)
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // Protobuf, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Protobuf serializer options.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct ProtobufSerializerOptions {
+    /// The path to the protobuf descriptor set file.
+    ///
+    /// This file is the output of `protoc -o <path> ...`
+    #[configurable(metadata(docs::examples = "/etc/vector/protobuf_descriptor_set.desc"))]
+    pub descriptor_set_path: PathBuf,
```

Comment:
🥜 (nit) The encoder and decoder options are identical. However the fields have different names `descriptor_set_path` vs `desc_file`. Let's pick one. Feel free to skip this refactoring.

Reply:
Picked `desc_file` for backwards compatibility.